base/abstractarray.jl

```diff
@@ -1481,12 +1481,11 @@ vcat(V::AbstractVector{T}...) where {T} = typed_vcat(T, V...)
 # but that solution currently fails (see #27188 and #27224)
 AbstractVecOrTuple{T} = Union{AbstractVector{<:T}, Tuple{Vararg{T}}}
 
-function _typed_vcat(::Type{T}, V::AbstractVecOrTuple{AbstractVector}) where T
-    n = 0
-    for Vk in V
-        n += Int(length(Vk))::Int
-    end
-    a = similar(V[1], T, n)
+_typed_vcat_similar(V, T, n) = similar(V[1], T, n)
+_typed_vcat(::Type{T}, V::AbstractVecOrTuple{AbstractVector}) where T =
+    _typed_vcat!(_typed_vcat_similar(V, T, mapreduce(length, +, V)), V)
```

Comment:
From the perspective of Base & stdlibs, this change seems safe because `_typed_vcat` has no backedges that extend deeper than `typed_vcat`---we expunged them from the system image (probably mostly in #37163). You can create more backedges with `vcat(1:3, [1,2,3])`, and so this may cause some worsening of package interactions, but I think that can argued to be their problem rather than Julia's.